### PR TITLE
Introduce the preview feature attribute

### PR DESCRIPTION
**Pull request by pgovind**
Unblocks some PRs and contributes to https://github.com/dotnet/designs/blob/main/accepted/2021/preview-features/preview-features.md

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Versioning/RequiresPreviewFeaturesAttribute.cs
Comment (jeffhandley) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Versioning/RequiresPreviewFeaturesAttribute.cs:15]: I know the design doc has `AllowMultiple = true`, but I'm not sure there's any reason to have that be `true`.

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (jkotas) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:10343]: This does not look like the auto-generated source. You may want to follow https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/updating-ref-source.md to avoid introducing delta with auto-generated source. (If there are any unnecessary delta that sneaked in in the meantime, feel free to include them in this PR as well.)

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (jkotas) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:10348]: Add a basic code-coverage test? (Look for e.g. CallerFilePathAttributeTests to what I have in mind.)

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (pgovind) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:10343]: Hmm I think there is something missing here. I don't understand the process here too well though. To update the ref in System.Runtime, the type forward needs to exist in the src? Anyway, to get unblocked here, I create a local file in System.Runtime with this attribute, built it with GenerateReferenceSource to auto-generate the changes for System.Runtime.cs and then deleted the local file. @safern, @ericstj maybe there's a better way? 

Also, is System.Runtime a full facade now?

**Review by jeffhandley — APPROVED** [review]

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (tannergooding) [reply src/libraries/System.Runtime/ref/System.Runtime.cs]: `System.Runtime` has always been problematic to regenerate references for. I think about half the time I try it works and the other half I ping Santi that its not working again 😄

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:858]: Nit: It would be nice to put this into the existing block of System\Runtime\... files

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Versioning/RequiresPreviewFeaturesAttribute.cs
Comment (pgovind) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Versioning/RequiresPreviewFeaturesAttribute.cs:16]: I asked Immo that exact question earlier today. I wasn't sure if we'd been through an API review already, so I chose to be conservative.

@terrajobst

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (safern) [reply src/libraries/System.Runtime/ref/System.Runtime.cs]: > Also, is System.Runtime a full facade now?

It has been like that for a long time already. `GenerateReferenceSource` should work fine if the type is defined in System.Private.CoreLib, make sure to run it with `--no-incremental` when invoking `dotnet build /t:GenerateReferenceSource`, feel free to ping me if you can't get this working.

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:858]: This comment was meant to be in the tests .csproj file.

File: src/libraries/System.Runtime/tests/System.Runtime.Tests.csproj
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests.csproj:122]: (attaching the comment to the right line)

Nit: It would be nice to put this into the existing block of System\Runtime... files

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Versioning/RequiresPreviewFeaturesAttribute.cs
Comment (pgovind) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Versioning/RequiresPreviewFeaturesAttribute.cs:16]: `Interface` and `Delegate` were added to the spec just now with https://github.com/dotnet/designs/pull/218/files. Will update this PR shortly

**Comment by pgovind** [comment]
Alright I just updated the PR to include `Delegate` and `Interface`. Will wait for another approval before I hit merge.

**Review by tannergooding — APPROVED** [review]

**Pull request merged**
